### ISSUE TITLE
Bad Loss History, Callback vs. Progressbar Loss Mismatch

### ISSUE THREAD

**Issue by QuiteAFoxtrot** [tensorflow/tensorflow#43470]
**System information**
- Have I written custom code - Yes
- OS Platform and Distribution - RHEL7
- Mobile device - None
- TensorFlow installed from - pip
- TensorFlow version v2.3.0-rc2-23-gb36436b087 2.3.0
- Python version: 3.7
- CUDA/cuDNN version: 
cudatoolkit               10.1.243             h6bb024c_0  
cudnn                     7.6.5                cuda10.1_0  
- GPU model and memory: 4x Nvidia TITAN X (Maxwell)

**Describe the current behavior**

I am working with a GAN, using code by subclassing tf.keras.Model in the style of: https://keras.io/examples/generative/dcgan_overriding_train_step/

I attempted to make a plot of the generator loss and the discriminator loss using the history object from the result of model.fit(), and noticed the values in the history object were all constants, despite obvious variability as displayed by the progress bar.  I dug deeper, and added a callback to print the `d_loss` and `g_loss` after `on_epoch_end` and `on_train_batch_end`, switched the model.compile() to include `run_eagerly=True`, added a print statement at the end of `train_step()` and I noticed that they do not match the values being printed by the progress bar.

Example terminal output:
```
Epoch 00001: LearningRateScheduler reducing learning rate to 0.01.
Epoch 1/4

Train Step:  tf.Tensor(0.72807, shape=(), dtype=float32) tf.Tensor(0.46352494, shape=(), dtype=float32)
Batch End:  {'d_loss': 0.7280700206756592, 'g_loss': 0.4635249376296997}
  1/100 [..............................] - ETA: 0s - d_loss: 0.7281 - g_loss: 0.4635

Train Step:  tf.Tensor(0.7460346, shape=(), dtype=float32) tf.Tensor(0.36776978, shape=(), dtype=float32)
Batch End:  {'d_loss': 0.7460346221923828, 'g_loss': 0.3677697777748108}
  2/100 [..............................] - ETA: 12s - d_loss: 0.7371 - g_loss: 0.4156

Train Step:  tf.Tensor(0.67379105, shape=(), dtype=float32) tf.Tensor(0.6921332, shape=(), dtype=float32)
Batch End:  {'d_loss': 0.6737910509109497, 'g_loss': 0.6921331882476807}
  3/100 [..............................] - ETA: 15s - d_loss: 0.7160 - g_loss: 0.5078
```

As you can see, on the first batch, all the values match precisely - the print statement at the end of `train_step`, the callback for `on_train_batch_end` and the progressbar output all match.  By the 2nd batch, `train_step` and `on_train_batch_end` are reporting a value of 0.746 for d_loss by the progressbar is printing a value of .7371, and g_loss is .3677 vs .4156.  

Originally, I suspected the distribute_strategy might be involved, so I commented that out and I re-ran with `CUDA_VISIBLE_DEVICES=0`, and got the same value mismatching.  I have this issue regardless of whether I run with `run_eagerly=True` or False.

My actual `train_step` for reference in case I'm just doing something stupid:
```
import tensorflow as tf
from tensorflow.nn import sigmoid_cross_entropy_with_logits

def train_step(self, train_step_data):
        
        high_res_imgs = train_step_data[1]
        
        low_res_imgs = train_step_data[0] 
        
        with tf.GradientTape(persistent=True) as tape:
           
            generated_images = self.generator(low_res_imgs)
    
            logits_on_generated = self.discriminator(generated_images)
            
            logits_on_real = self.discriminator(high_res_imgs)
            
            d_real_loss = sigmoid_cross_entropy_with_logits(labels=tf.ones_like(logits_on_real), logits=logits_on_real)
            
            d_gen_loss = sigmoid_cross_entropy_with_logits(labels=tf.zeros_like(logits_on_generated), logits=logits_on_generated)
            
            d_loss = 0.5*(d_real_loss + d_gen_loss)
            
            
            g_adv_loss = sigmoid_cross_entropy_with_logits(labels=tf.ones_like(logits_on_generated), logits=logits_on_generated)
            
            
            g_loss = g_adv_loss


        ################################################################################
        # Now backpropagate discriminator based on the losses
        ################################################################################
        grads = tape.gradient(d_loss, self.discriminator.trainable_weights)
        
        self.d_optimizer.apply_gradients(zip(grads, self.discriminator.trainable_weights))
        
        ################################################################################
        # Now backpropagate generator based on the losses
        ################################################################################
        grads = tape.gradient(g_loss, self.generator.trainable_weights)
        
        self.g_optimizer.apply_gradients(zip(grads, self.generator.trainable_weights))


        #print(d_loss.shape)
        #print(d_loss.numpy())
        print("\nTrain Step: ", d_loss[0,0], g_loss[0,0])
        return {"d_loss": d_loss[0,0], "g_loss": g_loss[0,0]}
```

I still have this issue regardless of whether I just return d_loss and g_loss or 'd_loss[0,0]' and g_loss[0,0].

And my callback:
```
class ModelPrinter(tf.keras.callbacks.Callback):
        
        def __init__(self, model):
            
            self.model = model
            
        def on_epoch_end(self, epoch, logs=None):
            
            print(self.model)
            print("\nEpoch End: ", logs)
            
        def on_train_batch_end(self, step_num, logs=None):
            
            print("\nBatch End: ", logs)
```

**Comment by bhack** [comment]
Can you have something very, very minimal code snippet that we could copy paste and run to reproduce this. If you have dummy input it is better so we don't need to copy files.

Also if you can share a minimal running Colab could be useful.

**Comment by QuiteAFoxtrot** [comment]
Self-contained example script:
```
import tensorflow as tf

from tensorflow.nn import sigmoid_cross_entropy_with_logits
from tensorflow.keras.losses import mean_squared_error
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.applications import VGG19

from tensorflow.keras.layers import Dense, \
                                    Activation, \
                                    Input, \
                                    Flatten, \
                                    Dropout, \
                                    LSTM, \
                                    Conv2D, \
                                    Convolution2DTranspose, \
                                    UpSampling2D, \
                                    Reshape, \
                                    BatchNormalization, \
                                    MaxPooling2D, \
                                    PReLU, \
                                    LeakyReLU, \
                                    ReLU, \
                                    Add
                                    
from tensorflow.keras.models import Sequential, Model

import numpy

        
        

def create_generator_CloudGAN(input_shape):
    
    gen_input = Input(shape=input_shape, name="Generator_Input_Layer")
    
    model = Conv2D(filters=256, 
                   kernel_size=9, 
                   strides=1, 
                   padding="same")(gen_input)
    
    model = PReLU(alpha_initializer='zeros', 
                  alpha_regularizer=None, 
                  alpha_constraint=None, 
                  shared_axes=[1,2])(model)
    
    model = Convolution2DTranspose(filters=64,
                                   kernel_size=3,
                                   strides=2,
                                   padding='same')(model)
    model = BatchNormalization()(model)
    
    model = ReLU()(model)
    
    
                                   
    model = Convolution2DTranspose(filters=16,
                                   kernel_size=3,
                                   strides=2,
                                   padding='same')(model)
    model = BatchNormalization()(model)
    
    model = ReLU()(model)
    
    
                                   
    model = Convolution2DTranspose(filters=4,
                                   kernel_size=3,
                                   strides=2,
                                   padding='same')(model)
    model = BatchNormalization()(model)
    
    model = ReLU()(model)
    
    
                                   
    model = Convolution2DTranspose(filters=1,
                                   kernel_size=3,
                                   strides=2,
                                   padding='same')(model)
    model = BatchNormalization()(model)
    
    model = ReLU()(model)
    
    generator_model = Model(inputs=gen_input, outputs=model)
    
    return generator_model





def discriminator_block(model, filters, kernel_size, strides):
    
    model = Conv2D(filters=filters, 
                   kernel_size=kernel_size, 
                   strides=strides, 
                   padding="same")(model)
                   
    model = BatchNormalization(momentum = 0.5)(model)
    
    model = LeakyReLU(alpha = 0.2)(model)
    
    return model
    

def create_discriminator(image_shape):
    
        
    input = Input(shape=image_shape, name="Discriminator_Input_Layer")
    
    model = Conv2D(filters=64, 
                   kernel_size=3, 
                   strides=1, 
                   padding="same")(input)
    
    model = LeakyReLU(alpha = 0.2)(model)
    
    model = discriminator_block(model, 64, 3, 2)
    model = discriminator_block(model, 64, 3, 2)
    model = discriminator_block(model, 64, 3, 2)
    
    
    model = Flatten()(model)
    model = Dense(256)(model)
    model = LeakyReLU(alpha = 0.2)(model)
   
    model = Dense(1)(model)
    model = Activation('sigmoid')(model) 
    
    discriminator_model = Model(inputs=input, outputs=model)
    
    return discriminator_model





class TestGAN(tf.keras.Model):
    
    def __init__(self,
                 high_res_shape,
                 low_res_shape,
                 discrim_lr, 
                 gen_lr):
        
        super(TestGAN, self).__init__()
        
        self.noise_shape = low_res_shape
        
        #self.output_shape = high_res_shape
        
        self.discriminator = create_discriminator(high_res_shape)
        
        self.generator = create_generator_CloudGAN(low_res_shape)
        
        self.d_optimizer = Adam(learning_rate=discrim_lr)
        
        self.g_optimizer = Adam(learning_rate=gen_lr)
        
        
    
    def train_step(self, train_step_data):
        
        low_res_imgs = train_step_data[0] 

        high_res_imgs = train_step_data[1]
        
        #Need persistent=True because we call the tape.gradient() twice
        with tf.GradientTape(persistent=True) as tape:
            
            generated_images = self.generator(low_res_imgs)
    
            logits_on_generated = self.discriminator(generated_images)
            
            logits_on_real = self.discriminator(high_res_imgs)
            
            
            
            ################################################################################
            
            real_labels = tf.ones_like(logits_on_real)
            real_labels += 0.05 * tf.random.uniform(tf.shape(real_labels))
            
            generated_labels = tf.zeros_like(logits_on_generated)
            generated_labels += 0.05 * tf.random.uniform(tf.shape(generated_labels))
    
            d_real_loss = sigmoid_cross_entropy_with_logits(labels=real_labels, logits=logits_on_real)
            
            d_gen_loss = sigmoid_cross_entropy_with_logits(labels=generated_labels, logits=logits_on_generated)
            
            d_loss = 0.5*(d_real_loss + d_gen_loss)
            
            
            
            g_loss = sigmoid_cross_entropy_with_logits(labels=tf.ones_like(logits_on_generated), logits=logits_on_generated)
            
            
        ################################################################################
        grads = tape.gradient(d_loss, self.discriminator.trainable_weights)
        
        self.d_optimizer.apply_gradients(zip(grads, self.discriminator.trainable_weights))
        
        ################################################################################
        grads = tape.gradient(g_loss, self.generator.trainable_weights)
        
        self.g_optimizer.apply_gradients(zip(grads, self.generator.trainable_weights))


        #print(d_loss.shape)
        #print(d_loss.numpy())
        print("\nTrain Step: ", d_loss[0,0], g_loss[0,0])
        return {"d_loss": d_loss[0,0], "g_loss": g_loss[0,0]}
    
    
    
    def call(self, inputs):
        
        print("Call, inputs: ", inputs.shape)
        
        return self.generator(inputs)
    

class ModelPrinter(tf.keras.callbacks.Callback):
        
    def __init__(self, model):
    
        self.model = model
    
    def on_epoch_end(self, epoch, logs=None):
    
        print("\nEpoch End: ", logs)
    
    def on_train_batch_end(self, step_num, logs=None):
    
        print("\nBatch End: ", logs)


class FakeDataGenerator():
    def __init__(self, batch_size, high_res_shape, low_res_shape):
        self.batch_size = batch_size
        self.high_res_shape = high_res_shape
        self.low_res_shape = low_res_shape

    def __iter__(self):
        return self

    def next(self):
        
        return self.__next__()

    def __next__(self):

        high_res = numpy.random.rand(self.batch_size, 
                                     self.high_res_shape[0],
                                     self.high_res_shape[1],
                                     1)

        low_res = numpy.random.rand(self.batch_size,
                                    self.low_res_shape[0],
                                    self.low_res_shape[1],
                                    1)

        return low_res, high_res



if __name__ == "__main__":


    batch_size = 4

    high_res_shape = (256, 256, 1)

    low_res_shape = (16, 16, 1)

    data_gen = FakeDataGenerator(batch_size,
                                 high_res_shape,
                                 low_res_shape)

    model = TestGAN(high_res_shape,
                    low_res_shape,
                    1.0e-4,
                    1.0e-4)

    model.compile(run_eagerly=True)


    callbacks_list = []

    callbacks_list.append(ModelPrinter(model))
        
    history = model.fit(data_gen, 
                            steps_per_epoch=100, 
                            epochs=3,
                            callbacks=callbacks_list, 
                            initial_epoch=0, 
                            verbose=1)

```

Same problem in output (note I added some newlines in the output below to make the trains steps easier to separate visually):
```
Epoch 1/3

Train Step:  tf.Tensor(0.70294505, shape=(), dtype=float32) tf.Tensor(0.47407043, shape=(), dtype=float32)
Batch End:  {'d_loss': 0.7029450535774231, 'g_loss': 0.4740704298019409}
  1/100 [..............................] - ETA: 0s - d_loss: 0.7029 - g_loss: 0.4741

Train Step:  tf.Tensor(0.6781775, shape=(), dtype=float32) tf.Tensor(0.4736327, shape=(), dtype=float32)
Batch End:  {'d_loss': 0.6781774759292603, 'g_loss': 0.47363269329071045}
  2/100 [..............................] - ETA: 16s - d_loss: 0.6906 - g_loss: 0.4739

Train Step:  tf.Tensor(0.65583074, shape=(), dtype=float32) tf.Tensor(0.47412047, shape=(), dtype=float32)
Batch End:  {'d_loss': 0.6558307409286499, 'g_loss': 0.47412046790122986}
  3/100 [..............................] - ETA: 21s - d_loss: 0.6790 - g_loss: 0.4739

```

**Comment by bhack** [comment]
Cause the progress bar logic is different.
If you try to insert a `print(values)` in the progressbar update method you will see the same values of in your Train Step/Batch End:
https://github.com/tensorflow/tensorflow/blob/f3462c311d2b5839a59f155ad82d3f3b3759da61/tensorflow/python/keras/utils/generic_utils.py#L530-L547

Then follow the code to see how `info` is composed.

**Comment by QuiteAFoxtrot** [comment]
Oh I think I see
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/utils/generic_utils.py#L650

It looks like its displaying a running mean instead of the raw value.  Maybe this just needs to be updated to a documentation update somewhere?

**Comment by bhack** [comment]
It Is not so explicit but It is described in params at https://www.tensorflow.org/api_docs/python/tf/keras/utils/Progbar

If you want you can close this and open a documentation improvement ISSUE or directly a PR on Doc
